compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -1547,50 +1315,72 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
             val incomingEdgeKind = incomingEdges.getValue(previousNode).kind
-
-            if (isDead && incomingEdgeKind.usedInDeadDfa) {
-                previousFlows += previousNode.flow
-            } else if (incomingEdgeKind.usedInDfa) {
-                previousFlows += previousNode.flow
-            }
-
             if (incomingEdgeKind == EdgeKind.DeadForward) {
                 deadForwardCount++
             }
+            if (incomingEdgeKind.usedInDfa || (isDead && incomingEdgeKind.usedInDeadDfa)) {
+                previousFlows += previousNode.flow
+            }
+        }
+        return logicSystem.joinFlow(previousFlows).also {
+            flow = it
+            if (updateReceivers || previousFlows.size + deadForwardCount > 1) {
+                resetReceivers(it)
+            }
         }
-        return logicSystem.joinFlow(previousFlows) to (previousFlows.size + deadForwardCount)
     }
 
-    private fun <T : CFGNode<*>> T.mergeIncomingFlow(
-        // This flag should be set true if we're changing flow branches from one to another (e.g. in when, try->catch)
-        updateReceivers: Boolean = false,
-        shouldForkFlow: Boolean = false
-    ): T = this.also { node ->
-        var (flow, incomingEdges) = computeIncomingFlow()
-        if (updateReceivers || incomingEdges > 1) {
-            logicSystem.updateAllReceivers(flow)
-        }
-        if (shouldForkFlow) {
-            flow = flow.fork()
+    private fun resetReceivers(flow: FLOW) {
+        receiverStack.forEach {
+            variableStorage.getRealVariable(flow, it.boundSymbol, it.receiverExpression)?.let { variable ->
+                receiverUpdated(flow, variable)
+            }
         }
-        node.flow = flow
     }
 
     private fun FLOW.addImplication(statement: Implication) {
+        val effect = statement.effect
+        if (effect is OperationStatement) {
+            val variable = effect.variable
+            if (variable.isReal()) {
+                when (effect.operation) {
+                    Operation.EqNull -> variable typeEq nullableNothing
+                    Operation.NotEqNull -> variable typeEq any
+                    else -> null
+                }?.let { logicSystem.addImplication(this, statement.condition implies it) }
+            }
+        }
         logicSystem.addImplication(this, statement)
     }
 
     private fun FLOW.addTypeStatement(info: TypeStatement) {
         logicSystem.addTypeStatement(this, info)
+        if (info.variable.isThisReference) {
+            receiverUpdated(this, info.variable)
+        }
     }
 
-    private fun FLOW.fork(): FLOW {
-        return logicSystem.forkFlow(this)
+    private fun FLOW.fork(): FLOW =
+        logicSystem.forkFlow(this)
+
+    private fun FLOW.commitOperationStatement(statement: OperationStatement, shouldRemoveSynthetics: Boolean) {
+        logicSystem.approveOperationStatement(this, statement, shouldRemoveSynthetics).values.forEach {
```

Comment:
It seems like with all the new flow forks `shouldRemoveSynthetics` can just always be true, but currently this causes some problems with `if(someCall()!!) { ... }` because `!!` does not rewrite statements on `someCall()` to be about the entire expression (so if there is a statement like `someCall() == true => ...`, it gets lost).